ocaml/runtime4/obj.c

```diff
@@ -27,22 +28,34 @@
 #include "caml/minor_gc.h"
 #include "caml/misc.h"
 #include "caml/mlvalues.h"
+#include "caml/platform.h"
 #include "caml/prims.h"
 #include "caml/signals.h"
 
-CAMLprim value caml_obj_tag(value arg)
+static int obj_tag (value arg)
 {
+  header_t hd;
+
   if (Is_long (arg)){
-    return Val_int (1000);   /* int_tag */
+    return 1000;   /* int_tag */
   }else if ((long) arg & (sizeof (value) - 1)){
-    return Val_int (1002);   /* unaligned_tag */
+    return 1002;   /* unaligned_tag */
   }else if (Is_in_value_area (arg)){
-    return Val_int(Tag_val(arg));
+    /* The acquire load ensures that reading the field of a Forward_tag
+       block in stdlib/camlinternalLazy.ml:force_gen has the necessary
+       synchronization. */
+    hd = (header_t)atomic_load_acquire(Hp_atomic_val(arg));
```

Comment:
I'm not sure the atomic_load here is necessary, since we have a single domain at all times, and thus there should be no synchronization needed. Maybe someone more knowledgeable about this could give their opinion?

Reply:
This sounds correct to me.